server/src/main/java/org/cloudfoundry/identity/uaa/provider/saml/SamlIdentityProviderConfigurator.java

```diff
@@ -44,7 +47,13 @@ public List<SamlIdentityProviderDefinition> getIdentityProviderDefinitions() {
 
     public List<SamlIdentityProviderDefinition> getIdentityProviderDefinitionsForZone(IdentityZone zone) {
         List<SamlIdentityProviderDefinition> result = new LinkedList<>();
-        for (IdentityProvider provider : providerProvisioning.retrieveActive(zone.getId())) {
+        String zoneId = zone.getId();
+        List<IdentityProvider> idpsForZone = requestScopedIdpDefinitionsForZone.getIdps(zoneId);
+        if (idpsForZone == null) {
+            idpsForZone = providerProvisioning.retrieveActive(zoneId);
```

Comment:
Looks right and simple enough, but you are explicitly implementing cache...  Have you considered using @Cachable annotation for retrieveAll() in JdbcIdentityProviderProvisioning and using spring provided ConcurrentMapCacheManager instead?  You could have per-request cache manager

@Cacheable(cacheManager = "requestScopedIdpDefinitionsCacheManager"...)
public List<IdentityProvider> retrieveAll(boolean activeOnly, String zoneId) {
...
}

@RequestScope
@Component("requestScopedIdpDefinitionsCacheManager")
public CacheManager requestScopedIdpDefinitionsCacheManager() {
    return new ConcurrentMapCacheManager();
}

https://docs.spring.io/spring-framework/docs/5.3.18/javadoc-api/org/springframework/cache/concurrent/ConcurrentMapCacheManager.html

If it works, it will simplify code here, no need for if and explicit get from cache...

Reply:
I tried using Spring's cache as described in https://rcarrillo.dev/request-scoped-caching-with-spring/ but it somehow doesn't work (test failed and I do see unnecessary queries as before without the optimization). See https://github.com/GESoftware-CF/uaa/commit/5540261360b9c7e5ef89a112241ab0c39151861e